module/zfs/dbuf.c

```diff
@@ -2638,28 +2638,28 @@ dmu_buf_will_dirty_impl(dmu_buf_t *db_fake, int flags, dmu_tx_t *tx)
 	mutex_enter(&db->db_mtx);
 
 	if (db->db_state == DB_CACHED || db->db_state == DB_NOFILL) {
-		dbuf_dirty_record_t *dr = dbuf_find_dirty_eq(db, tx->tx_txg);
 		/*
 		 * It's possible that it is already dirty but not cached,
 		 * because there are some calls to dbuf_dirty() that don't
 		 * go through dmu_buf_will_dirty().
 		 */
-		if (dr != NULL) {
-			if (dr->dt.dl.dr_brtwrite) {
-				/*
-				 * Block cloning: If we are dirtying a cloned
-				 * block, we cannot simply redirty it, because
-				 * this dr has no data associated with it.
-				 * We will go through a full undirtying below,
-				 * before dirtying it again.
-				 */
-				undirty = B_TRUE;
-			} else {
-				/* This dbuf is already dirty and cached. */
-				dbuf_redirty(dr);
-				mutex_exit(&db->db_mtx);
-				return;
-			}
+		dbuf_dirty_record_t *dr = dbuf_find_dirty_eq(db, tx->tx_txg);
+		if (dr != NULL && dr->dr_dbuf != NULL &&
+		    dr->dr_dbuf->db_level == 0 &&
```

Comment:
I drew quick inspiration from [`dbuf_sync_list`](https://github.com/openzfs/zfs/blob/master/module/zfs/dbuf.c#L4693) to exclude both 2 of the 3 `dr` types, just in case. You're most likely right, the tripped VERIFY0 in the panics would suggest that only db_level != 0 is the problem.

Reply:
Still, in case the semantics subtly change in the future and `dmu_buf_will_dirty_impl` somehow gets called in the wrong situation in the `dll` case, we have a potential NULL pointer deref here. I'd vote to keep the `dr->dr_dbuf != NULL` part in there if it's not too much trouble/too misleading.

Reply:
As I understand, the whole idea of lightweight write is to not allocate dbufs.  It should be physically impossible to get into dmu_buf_will_dirty_impl() for it, since needs dbuf on input.

Reply:
ok, this makes the change a bit simpler then :)